### PR TITLE
TN-1569 WIP: DON'T MERGE YET

### PR DESCRIPTION
**Pull request by pbugni**
Work in progress for questionnaire bank timeline.

Getting close!  Now includes a view to see any patient's timeline at:

`/api/patient/<patient_id>/timeline`

### CODE REVIEW

File: portal/models/qb_timeline.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    @property
+    def status(self):
+        return self._status
+
+    @status.setter
+    def status(self, value):
+        self._status = getattr(OverallStatus, value).name
+
+
+system_trigger = datetime(2015, 1, 1, 12, 0, 0)
```
Comment (ivan-c) [inline portal/models/qb_timeline.py]: Maybe something further back and standardized, like the Unix Epoch?

File: portal/models/questionnaire_bank.py [hunk 2]
```diff
@@ -638,6 +685,23 @@ def add_if_not_found(self, commit_immediately=False):
         return self
 
 
+@dogpile_cache.region('qb_query_cache')
```
Comment (ivan-c) [inline portal/models/questionnaire_bank.py]: Are you caching DB objects? I know we've had issues with stale DB objects with celery

File: portal/models/qb_timeline.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+from ..database import db
+from ..date_tools import RelativeDelta
+from .assessment_status import OverallStatus
+from .questionnaire_bank import QuestionnaireBank, qbs_by_rp
+from .questionnaire_response import QNR_results
+from .user_consent import consent_withdrawal_dates
+
+
+class QBT(db.Model):
```
[the remark this answers is not in the document]
Comment (pbugni) [reply portal/models/qb_timeline.py]: dogpile doesn't have the same granular controls we get from SQL.  plus, we treat redis as a volatile cache - this really will alleviate all the work to build out users data other than when submissions come in, and then just the one user at a time.

File: portal/models/qb_timeline.py (hunk 1 above)
Comment (pbugni) [reply portal/models/qb_timeline.py]: we ran into trouble with dates < 1900 w/ one of the python date libraries.  can go to that if you'd rather?  (jan 1, 1900)

File: portal/models/questionnaire_bank.py (hunk 2 above)
Comment (pbugni) [reply portal/models/questionnaire_bank.py]: yes - and i believe i found the workaround - not the updated comment and all clients refresh the cached objects in the session.

Comment (pbugni) [reply portal/models/questionnaire_bank.py]: I am finding the *win* by caching these is minimal (like 15% over the entire run) - tempted to remove this caching as it does add complications.

File: portal/models/qb_timeline.py (hunk 1 above)
Comment (pbugni) [reply portal/models/qb_timeline.py]: killed this approach after finding it dysfunctional w/ the "month math" we require.

File: portal/models/questionnaire_bank.py (hunk 2 above)
Comment (pbugni) [reply portal/models/questionnaire_bank.py]: I *did* remove this caching - premature optimization, more trouble to get right than the minor savings, which we only see on the *first* run.

**Pull request closed without merging**